### PR TITLE
8265899: Use pattern matching for instanceof at module jdk.compiler(part 1)

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This series of patches update the code of the module jdk.compiler by using the pattern matching for instanceof.

This patch would revise the following packages:
```
com.sun.tools.javac.api
com.sun.tools.javac.file
com.sun.tools.javac.main
com.sun.tools.javac.model
com.sun.tools.javac.platform
com.sun.tools.javac.tree
com.sun.tools.javac.util
```

Thank you for taking the time to review.

--
Best Regards,
Guoxiong.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265899](https://bugs.openjdk.java.net/browse/JDK-8265899): Use pattern matching for instanceof at module jdk.compiler(part 1)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3673/head:pull/3673` \
`$ git checkout pull/3673`

Update a local copy of the PR: \
`$ git checkout pull/3673` \
`$ git pull https://git.openjdk.java.net/jdk pull/3673/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3673`

View PR using the GUI difftool: \
`$ git pr show -t 3673`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3673.diff">https://git.openjdk.java.net/jdk/pull/3673.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacScope.java [hunk 1]
```diff
@@ -127,10 +127,9 @@ public boolean isStarImportScope() {
     }
 
     public boolean equals(Object other) {
-        if (other instanceof JavacScope) {
-            JavacScope s = (JavacScope) other;
-            return (env.equals(s.env)
-                && isStarImportScope() == s.isStarImportScope());
+        if (other instanceof JavacScope javacScope) {
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacScope.java:130]: Can this be:

```
return other instanceof JavacScope javacScope &&
            env.equals(javacScope.env) &&
            isStarImportScope() == javacScope.isStarImportScope();
```

File: src/jdk.compiler/share/classes/com/sun/tools/javac/file/JavacFileManager.java [hunk 2]
```diff
@@ -957,17 +957,16 @@ private JavaFileObject getFileForOutput(Location location,
...
-    public Iterable<? extends JavaFileObject> getJavaFileObjectsFromPaths(
-        Collection<? extends Path> paths)
-    {
+    public Iterable<? extends JavaFileObject> getJavaFileObjectsFromPaths(Collection<? extends Path> paths) {
         ArrayList<PathFileObject> result;
-        if (paths instanceof Collection<?>)
-            result = new ArrayList<>(((Collection<?>)paths).size());
-        else
+        if (paths != null) {
+            result = new ArrayList<>(paths.size());
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/file/JavacFileManager.java:963]: This seems odd in the original code - I'd rather have somebody versed in the API double check this - @jonathan-gibbons  ? Anyway - this doesn't seem to be related with using pattern matching, so perhaps it's preferrable to just move it out of this PR.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/model/JavacTypes.java [hunk 3]
```diff
@@ -329,18 +329,17 @@ private void validateTypeNotIn(TypeMirror t, Set<TypeKind> invalidKinds) {
                 || elem.getModifiers().contains(Modifier.PRIVATE))
             return Collections.emptySet();
 
-        if (!(elem instanceof MethodSymbol))
+        if (!(elem instanceof MethodSymbol methodSymbol))
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/model/JavacTypes.java:332]: These two look odd - your cleanup is fine, but it seems in the first cast, we want to wrap with IllegalArgumentException instead of failing with ClassCastException. In the second cast (for class symbol) it seems we're ok with just CCE. I'm fine with leaving this as per your cleanup - just noting the inconsistency.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/JCTree.java [hunk 4]
```diff
@@ -2178,7 +2178,7 @@ protected JCInstanceOf(JCExpression expr, JCTree pattern) {
 
         @Override @DefinedBy(Api.COMPILER_TREE)
         public JCPattern getPattern() {
-            return pattern instanceof JCPattern ? (JCPattern) pattern : null;
+            return pattern instanceof JCPattern jcPattern ? jcPattern : null;
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/tree/JCTree.java:2181]: Using patterns to implement patterns support :-)

**Review by mcimadamore — COMMENTED** [review]
Very nice cleanup - left few comments

File: src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacScope.java (hunk 1 above)
Comment (lgxbslgx) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/api/JavacScope.java]: Fixed

File: src/jdk.compiler/share/classes/com/sun/tools/javac/file/JavacFileManager.java (hunk 2 above)
Comment (mcimadamore) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/file/JavacFileManager.java:963]: OK

**Review by mcimadamore — APPROVED** [review]
Looks good

**Comment by lgxbslgx** [comment]
@mcimadamore Thank you for your review. May I ask your help to sponsor this patch?
